### ISSUE TITLE
Add Natural Language API's for IOS

### ISSUE THREAD

**Issue by bharat-biradar** [bharat-biradar/Google-Ml-Kit-plugin#153]
To add any one or all of the remaining Natural Language API's for IOS
- [x] On Device Translation
- [x] Entity Extraction
- [x] Smart Reply

**Comment by skanderhamdi** [comment]
Hello,
I'm not familiar with Swift and creating plugins, as I needed on device translation for iOS, I've created a quick solution for:

- Download TranslateRemoteModel
- Delete TranslationRemoteModel
- Check if a model is downloaded
- Translate text

1. Add these lines to your Podfile
`
pod 'Firebase'
pod 'GoogleMLKit/MLKitCore', '2.2.0'
pod 'GoogleMLKit/Translate', '2.2.0'
`
2. run `pod install`
3. Inside your project, create one dart file for each one of the classes:
`
class GoogleMLKitNLPiOS
{
    GoogleMLKitNLPiOS._();
    static const MethodChannel channel = MethodChannel('google_mlkit_nlp_ios');
    static final GoogleMLKitNLPiOS instance = GoogleMLKitNLPiOS._();
    iOSLanguageTranslator languageTranslator(
        {required String fromLanguage, required String toLanguage}) => iOSLanguageTranslator(toLanguage: toLanguage, fromLanguage: fromLanguage);
    iOSLanguageModelManager modelManager() => iOSLanguageModelManager();
}
class iOSLanguageTranslator {
  iOSLanguageTranslator({required String fromLanguage, required String toLanguage})
      : _fromLanguage = fromLanguage,
        _toLanguage = toLanguage;
  final String _fromLanguage;
  final String _toLanguage;
  Future<String> translateText(String text) async {
    final String translatedText = await GoogleMLKitNLPiOS.channel.invokeMethod(
        'translateText',
        <String, dynamic>{
          'fromLanguage': _fromLanguage,
          'toLanguage': _toLanguage,
          'text': text
        }
    );
    return translatedText;
  }
}
class iOSLanguageModelManager {
  iOSLanguageModelManager();

  Future<bool> deleteModel(String toDelete) async {
    final bool status = await GoogleMLKitNLPiOS.channel.invokeMethod(
        'deleteModel', <String, dynamic>{'languageCode': toDelete});
    return status;
  }

  Future<bool> downloadModel(String toDownload) async {
    final bool status = await GoogleMLKitNLPiOS.channel.invokeMethod(
        'downloadModel', <String, dynamic>{'languageCode': toDownload});
    return status;
  }

  Future<bool> isModelDownloaded(String langCode) async {
    final bool status = await GoogleMLKitNLPiOS.channel.invokeMethod(
        'isModelDownloaded', <String, dynamic>{'languageCode': langCode});
    return status;
  }
}
`
4. Now, create a Swift file beside AppDelegate.swift, name it for example: `GoogleMLKitNLP.swift`

`
import Foundation
import MLKit

class GoogleMLKitNLP
{
    init(){}
    var translator: Translator!
    let locale = Locale.current
    lazy var allLanguages = TranslateLanguage.allLanguages().sorted {
      return locale.localizedString(forLanguageCode: $0.rawValue)!
          < locale.localizedString(forLanguageCode: $1.rawValue)!
    }
    func createModel(forLanguage: TranslateLanguage) -> TranslateRemoteModel {
        return TranslateRemoteModel.translateRemoteModel(language: forLanguage)
    }
    // To check if language model is downloaded or not by passing its language code
    func isLanguageModelDownloaded(call: FlutterMethodCall, result: FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>{
           let languageCode = args["languageCode"] as? String
           if let translateLanguage = allLanguages.filter({$0.rawValue == languageCode}).first{
               let model = self.createModel(forLanguage: translateLanguage)
               let modelManager = ModelManager.modelManager()
               result(Bool(modelManager.isModelDownloaded(model)))
           }else{
               result(FlutterError.init(code: "INVALID_LANG_CODE", message: "Invalide language code", details: nil))
           }
        } else {
            result(FlutterError.init(code: "NO_LANG_CODE", message: "No language code for check", details: nil))
        }
    }
    // To delete a model from local storage
    func deleteModel(call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>{
           let languageCode = args["languageCode"] as? String
           if let translateLanguage = allLanguages.filter({$0.rawValue == languageCode}).first{
               let model = self.createModel(forLanguage: translateLanguage)
               let modelManager = ModelManager.modelManager()
               modelManager.deleteDownloadedModel(model) { error in
                  guard error == nil else {
                      result(false)
                      return
                  }
                   result(true)
               }
           }else{
               result(FlutterError.init(code: "INVALID_LANG_CODE", message: "Invalide language code", details: nil))
           }
        } else {
            result(FlutterError.init(code: "NO_LANG_CODE", message: "No language code for check", details: nil))
        }
    }
    // To download a model to the local storage
    func downloadModel(call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>{
           let languageCode = args["languageCode"] as? String
           if let translateLanguage = allLanguages.filter({$0.rawValue == languageCode}).first{
               let model = self.createModel(forLanguage: translateLanguage)
               let conditions = ModelDownloadConditions(
                   allowsCellularAccess: true,
                   allowsBackgroundDownloading: true
               )
               ModelManager.modelManager().download(model,conditions: conditions)
               NotificationCenter.default.addObserver(
                   forName: .mlkitModelDownloadDidSucceed,
                   object: nil,
                   queue: nil
               ) { notification in
                   result(true)
                   return
               }
               NotificationCenter.default.addObserver(
                   forName: .mlkitModelDownloadDidFail,
                   object: nil,
                   queue: nil
               ) { notification in
                   result(false)
                   return
                }
           }else{
               result(FlutterError.init(code: "INVALID_LANG_CODE", message: "Invalide language code", details: nil))
           }
        } else {
            result(FlutterError.init(code: "NO_LANG_CODE", message: "No language code for check", details: nil))
        }
    }
    // For translating text
    // Input: input lang + text + output lang
    // Output: translted text from input to output
    func translateText(call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>{
           // Get data from arguments as String?
           let fromLanguageCode = args["fromLanguage"] as? String
           let toLanguageCode = args["toLanguage"] as? String
           let originalText = args["text"] as? String
           // Create instances TranslateLanguage of selected languages
           let inputLanguage = allLanguages.filter({$0.rawValue == fromLanguageCode}).first
           let outputLanguage = allLanguages.filter({$0.rawValue == toLanguageCode}).first
           // Initialize Translator
           let translatorOptions = TranslatorOptions(sourceLanguage: inputLanguage!, targetLanguage: outputLanguage!)
           self.translator = Translator.translator(options: translatorOptions)
           self.translator.translate(originalText ?? "") { translationResult, error in
               guard error == nil else {
                   result("")
                   return
               }
               result(translationResult)
               return
           }
        } else {
            result(FlutterError.init(code: "REQUIRED_DATA_MISSING", message: "Input language, output language and text to translate should be sent", details: nil))
        }
    }
}
`

5. In your AppDelegate.swift, insert these lines:

In the the imports section
`import Firebase`

In the beginning of `application` function, insert the following lines:

`
FirebaseApp.configure()
      let controller: FlutterViewController = window?.rootViewController as! FlutterViewController
      let nlp_mlkit_channel = FlutterMethodChannel(name:"google_mlkit_nlp_ios",binaryMessenger: controller.binaryMessenger)
      let googleMLKitNLP = GoogleMLKitNLP()
      nlp_mlkit_channel.setMethodCallHandler({
        (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
        switch call.method {
          case "isModelDownloaded":
            googleMLKitNLP.isLanguageModelDownloaded(call: call ,result: result)
          case "deleteModel":
            googleMLKitNLP.deleteModel(call: call ,result: result)
          case "downloadModel":
            googleMLKitNLP.downloadModel(call: call ,result: result)
          case "translateText":
            googleMLKitNLP.translateText(call: call ,result: result)
          default:
            result(FlutterMethodNotImplemented)
        }
      })
`

Sorry for putting the code here, i had never contributing in plugin development, hope this help for On-Device Translation in iOS.

Thank you

**Comment by skanderhamdi** [comment]
Hello,
I'm not familiar with Swift and creating plugins, as I needed on device translation for iOS, I've created a quick solution for:

- Download TranslateRemoteModel
- Delete TranslationRemoteModel
- Check if a model is downloaded
- Translate text

Add these lines to your Podfile

```
pod 'Firebase'
pod 'GoogleMLKit/MLKitCore', '2.2.0'
pod 'GoogleMLKit/Translate', '2.2.0'
```

run `pod install`

Inside your project, create one dart file for each one of the classes:

```
class GoogleMLKitNLPiOS
{
    GoogleMLKitNLPiOS.();
    static const MethodChannel channel = MethodChannel('google_mlkit_nlp_ios');
    static final GoogleMLKitNLPiOS instance = GoogleMLKitNLPiOS.();
    iOSLanguageTranslator languageTranslator(
    {required String fromLanguage, required String toLanguage}) => iOSLanguageTranslator(toLanguage: toLanguage, fromLanguage: fromLanguage);
    iOSLanguageModelManager modelManager() => iOSLanguageModelManager();
}
```

```
class iOSLanguageTranslator {
    iOSLanguageTranslator({required String fromLanguage, required String toLanguage})
    : _fromLanguage = fromLanguage,
    _toLanguage = toLanguage;

    final String _fromLanguage;
    final String _toLanguage;
    Future translateText(String text) async {
        final String translatedText = await GoogleMLKitNLPiOS.channel.invokeMethod(
            'translateText',
            <String, dynamic>{
            'fromLanguage': _fromLanguage,
            'toLanguage': _toLanguage,
            'text': text
            }
        );
        return translatedText;
    }
}
```
```
class iOSLanguageModelManager {
    iOSLanguageModelManager();
    Future deleteModel(String toDelete) async {
        final bool status = await GoogleMLKitNLPiOS.channel.invokeMethod(
        'deleteModel', <String, dynamic>{'languageCode': toDelete});
        return status;
    }
    
    Future downloadModel(String toDownload) async {
        final bool status = await GoogleMLKitNLPiOS.channel.invokeMethod(
        'downloadModel', <String, dynamic>{'languageCode': toDownload});
        return status;
    }

    Future isModelDownloaded(String langCode) async {
        final bool status = await GoogleMLKitNLPiOS.channel.invokeMethod(
        'isModelDownloaded', <String, dynamic>{'languageCode': langCode});
        return status;
    }
}

```

4. Now, create a Swift file beside AppDelegate.swift, name it for example:GoogleMLKitNLP.swift`

```
import Foundation
import MLKit

class GoogleMLKitNLP
{
      init(){}
      var translator: Translator!
      let locale = Locale.current
      lazy var allLanguages = TranslateLanguage.allLanguages().sorted {
          return locale.localizedString(forLanguageCode: $0.rawValue)!
          < locale.localizedString(forLanguageCode: $1.rawValue)!
      }
      func createModel(forLanguage: TranslateLanguage) -> TranslateRemoteModel {
           return TranslateRemoteModel.translateRemoteModel(language: forLanguage)
      }

      // To check if language model is downloaded or not by passing its language code
      func isLanguageModelDownloaded(call: FlutterMethodCall, result: FlutterResult) {
          if let args = call.arguments as? Dictionary<String, Any>{
                let languageCode = args["languageCode"] as? String
                if let translateLanguage = allLanguages.filter({$0.rawValue == languageCode}).first{
                        let model = self.createModel(forLanguage: translateLanguage)
                        let modelManager = ModelManager.modelManager()
                        result(Bool(modelManager.isModelDownloaded(model)))
                  }else{
                        result(FlutterError.init(code: "INVALID_LANG_CODE", message: "Invalide language code", details: nil))
                  }
          } else {
                 result(FlutterError.init(code: "NO_LANG_CODE", message: "No language code for check", details: nil))
          }
      }
      // To delete a model from local storage
      func deleteModel(call: FlutterMethodCall, result: @escaping FlutterResult) {
          if let args = call.arguments as? Dictionary<String, Any>{
             let languageCode = args["languageCode"] as? String
             if let translateLanguage = allLanguages.filter({$0.rawValue == languageCode}).first{
                 let model = self.createModel(forLanguage: translateLanguage)
                 let modelManager = ModelManager.modelManager()
                 modelManager.deleteDownloadedModel(model) { error in
                    guard error == nil else {
                        result(false)
                        return
                    }
                     result(true)
                 }
             }else{
                 result(FlutterError.init(code: "INVALID_LANG_CODE", message: "Invalide language code", details: nil))
             }
          } else {
              result(FlutterError.init(code: "NO_LANG_CODE", message: "No language code for check", details: nil))
          }
      }
      // To download a model to the local storage
      func downloadModel(call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>{
           let languageCode = args["languageCode"] as? String
           if let translateLanguage = allLanguages.filter({$0.rawValue == languageCode}).first{
               let model = self.createModel(forLanguage: translateLanguage)
               let conditions = ModelDownloadConditions(
                   allowsCellularAccess: true,
                   allowsBackgroundDownloading: true
               )
               ModelManager.modelManager().download(model,conditions: conditions)
               NotificationCenter.default.addObserver(
                   forName: .mlkitModelDownloadDidSucceed,
                   object: nil,
                   queue: nil
               ) { notification in
                   result(true)
                   return
               }
               NotificationCenter.default.addObserver(
                   forName: .mlkitModelDownloadDidFail,
                   object: nil,
                   queue: nil
               ) { notification in
                   result(false)
                   return
                }
           }else{
               result(FlutterError.init(code: "INVALID_LANG_CODE", message: "Invalide language code", details: nil))
           }
        } else {
            result(FlutterError.init(code: "NO_LANG_CODE", message: "No language code for check", details: nil))
        }
      }
      // For translating text
      // Input: input lang + text + output lang
      // Output: translted text from input to output
      func translateText(call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>{
           // Get data from arguments as String?
           let fromLanguageCode = args["fromLanguage"] as? String
           let toLanguageCode = args["toLanguage"] as? String
           let originalText = args["text"] as? String
           // Create instances TranslateLanguage of selected languages
           let inputLanguage = allLanguages.filter({$0.rawValue == fromLanguageCode}).first
           let outputLanguage = allLanguages.filter({$0.rawValue == toLanguageCode}).first
           // Initialize Translator
           let translatorOptions = TranslatorOptions(sourceLanguage: inputLanguage!, targetLanguage: outputLanguage!)
           self.translator = Translator.translator(options: translatorOptions)
           self.translator.translate(originalText ?? "") { translationResult, error in
               guard error == nil else {
                   result("")
                   return
               }
               result(translationResult)
               return
           }
        } else {
            result(FlutterError.init(code: "REQUIRED_DATA_MISSING", message: "Input language, output language and text to translate should be sent", details: nil))
        }
    }
}
```

In your AppDelegate.swift, insert these lines:
In the the imports section

```
import Firebase
```

In the beginning of application function, insert the following lines:

```
FirebaseApp.configure()
let controller: FlutterViewController = window?.rootViewController as! FlutterViewController
let nlp_mlkit_channel = FlutterMethodChannel(name:"google_mlkit_nlp_ios",binaryMessenger: controller.binaryMessenger)
let googleMLKitNLP = GoogleMLKitNLP()
nlp_mlkit_channel.setMethodCallHandler({
        (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
        switch call.method {
          case "isModelDownloaded":
            googleMLKitNLP.isLanguageModelDownloaded(call: call ,result: result)
          case "deleteModel":
            googleMLKitNLP.deleteModel(call: call ,result: result)
          case "downloadModel":
            googleMLKitNLP.downloadModel(call: call ,result: result)
          case "translateText":
            googleMLKitNLP.translateText(call: call ,result: result)
          default:
            result(FlutterMethodNotImplemented)
        }
})
```
Now, you can invoke `downloadModel` , `downloadModel` , `isModelDownloaded` and `translateText` via your Dart code.

Sorry for putting the code here, i had never contributing in plugin development, hope this help for On-Device Translation in iOS.

Thank you

**Comment by p4-k4** [comment]
Hi.

Will there be an implementation for the above from @skanderhamdi ? Or is there a reason why this is still open?

**Comment by bharat-biradar** [comment]
Hi @p4-k4 the api's have been implemented in the pr #215, it might take another week to merge and publish all the plugins.

**Comment by p4-k4** [comment]
@bharat-biradar That's awesome news, I look forward to it! Thanks a tonne mate.
